django/db/models/fields/files.py

```diff
@@ -295,20 +297,26 @@ def contribute_to_class(self, cls, name, **kwargs):
         setattr(cls, self.name, self.descriptor_class(self))
 
     def get_directory_name(self):
```

Comment:
You are right with this and 

> We likely do need to offer backwards-compatibility (or at least provide an upgrade guide) even though those methods aren't documented -- they seem to have widespread usage. Take a look at https://github.com/search?q=generate_filename%28self%2C+instance%2C+filename%29&ref=searchresults&type=Code&utf8=%E2%9C%93 to get a sense for what users are doing.

In order to offer 100% compatibility the implementation should rather delegate get_directory_name and get_filename to the storage and keep the generate_filename implementation at the FileField class, I think this would imply adding a third method at the storage class for the os.path.join operations sadly, so we would end up with 3 new methods at the storage class rather than just 1. What do you suggest? I will think more about this tomorrow and also handle the above stuff.